File: LeetCode/C#/50. Pow(x, n)/Solution.cs
public class Solution 
{
    public double res;
    public double MyPow(double x, int n) 
    {
        if (n<0)
        {
            x=1/x;
            n*=-1;
        }
        if (n==0) return 1.0;
        if (x==0) return 0.0;
        BetterPow(n-1,1,x,x,x);
        return res;
    }
    public void BetterPow(int powsToGo, int currentPow, double currentNumber, double basePow, double resultNumber)
    {
        if (currentPow<powsToGo)
        {
            resultNumber*=currentNumber;
            currentNumber*=currentNumber;
            BetterPow(powsToGo-currentPow,currentPow*2, currentNumber, basePow, resultNumber);
        }
        else if (powsToGo>0)
        {
            BetterPow(powsToGo-1,1,basePow,basePow,resultNumber*basePow);
        }
        else res=resultNumber;
    }
}

File: LeetCode/C++/50. Pow(x, n)/Solution.cpp
class Solution {
private:
    double res;
    double betterPow(double startNum, long leftN, long currentN, double currentRes, double currentMult)
    {
        if (leftN==1) return currentRes;
        return currentN>=leftN ? betterPow(startNum,leftN,1,currentRes,startNum) : betterPow(startNum,leftN-currentN,2*currentN,currentRes*currentMult,currentMult*currentMult);
    }
public:
    double myPow(double x, long n) {
        if (n<0)
        {
            n*=-1;
            x=1/x;
        }
        return n==0 ? 1 : betterPow(x,n,1,x,x);
    }
};